ASP.NET Core OpenID Connect

damienbod opened the pull request:
PR for part 1 of this issue: https://github.com/dotnet/AspNetCore.Docs/issues/33877

fixes https://github.com/dotnet/AspNetCore.Docs/issues/30647

FYI: @blowdart @JeremyLikness @kevinchalet @halter73 @guardrex @mikekistler @jennyf19

@Rick-Anderson  A lot of information in this and I left out some bits like how OAuth PAR works, refresh implementation and just links some of the other docs instead direct doc in the doc.

image is a drawio png.

what do you think? 

Greetings Damien

Rick-Anderson commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,319 @@
+
+By [Damien Bowden](https://github.com/damienbod)
+
+[View or download sample code](~/security/authentication/configure-oidc-web-authentication/sample/oidc-net8)
```
[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/security/authentication/configure-oidc-web-authentication/sample/oidc-net8)

Rick-Anderson commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
````diff
@@ -0,0 +1,319 @@
+});
+```
+
+See [Secure an ASP.NET Core Blazor Web App with OpenID Connect (OIDC)](xref:core/blazor/security/blazor-web-app-with-oidc) for details on the different OpenID Connect options.
````
```suggestion
See [Secure an ASP.NET Core Blazor Web App with OpenID Connect (OIDC)](xref:blazor/security/blazor-web-app-with-oidc) for details on the different OpenID Connect options.
```

damienbod replied:
fixed

Rick-Anderson commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,319 @@
+
+Refer to the following document:
+
+[Secure an ASP.NET Core Blazor Web App with OpenID Connect (OIDC)](xref:sblazor/security/blazor-web-app-oidc)
```
```suggestion
[Secure an ASP.NET Core Blazor Web App with OpenID Connect (OIDC)](xref:blazor/security/blazor-web-app-oidc)
```

damienbod replied:
fixed

Rick-Anderson reviewed (changes requested):
Can you accept these changes to fix the build problems? So far looks amazing but this will need security approval.
Amazing work.

damienbod commented:
I remove this from the toc so that the build error goes away:

´´´
          - name: Configure OIDC web authentication
            uid: security/authentication/configure-oidc-web-authentication
´´´

damienbod commented:
@Rick-Anderson Fixed the links, removed the toc stuff, still 1 error which, any idea how to fix?

Rick-Anderson commented:
email sent to principals Nov 5th. Subject: Configure OIDC web authentication

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,319 @@
+    var policy = new AuthorizationPolicyBuilder()
+        .RequireAuthenticatedUser()
+        .Build();
+    options.Filters.Add(new AuthorizeFilter(policy));
```
I think it would probably be better to demonstrate setting a [FallbackPolicy](https://learn.microsoft.com/aspnet/core/security/authorization/secure-data?view=aspnetcore-8.0#require-authenticated-users] over MVC filters for this purpose so it would also work if people add SignalR, Blazor, minimal APIs, etc... to the project.

damienbod replied:
fixed

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,319 @@
+.AddCookie()
+.AddOpenIdConnect(options =>
+{
+	configuration.GetSection("OpenIDConnectSettings").Bind(options);
+	
+	options.Authority = configuration["OpenIDConnectSettings:Authority"];
+	options.ClientId = configuration["OpenIDConnectSettings:ClientId"];
+	options.ClientSecret = configuration["OpenIDConnectSettings:ClientSecret"];
```
```suggestion
    var oidcConfig = configuration.GetSection("OpenIDConnectSettings");

	  options.Authority = oidcConfig["Authority"];
	  options.ClientId = oidcConfig["ClientId"];
	  options.ClientSecret = oidcConfig["ClientSecret"];
```

Isn't the call to `Bind(options)` redundant? I like explicitly setting the properties over just calling `Bind` to show which ones matter. And it doesn't require the reader to understand config binding to get a general idea of what's going on.

Also, I think we try to use 4 spaces for indentation everywhere.

damienbod replied:
fixed

damienbod commented:
> When I filed #33877, I was imagining a smaller sample that didn't pull in Microsoft.AspNetCore.Identity, EF, OpenIddict, etc... I appreciate that this sample gives the reader something where they can just hit F5 and see something working, but it does add a lot of complexity to the sample for someone who just wants to learn how to integrate their ASP.NET Core app with a non-self-hosted OIDC provider like Auth0 or Keycloak.
> 
> Would you be okay with removing the IdentityProvider portion of the sample and leaving just the RazorPageOidcClient part? I might rename RazorPageOidcClient to just RazorPagesOidc too. I think the "Client" part of it adds some clarity when you also include the OIDC provider in the sample, but it could create confusion for the majority of developers new to OIDC who just want to set up what they see as a ASP.NET Core "Server" to authenticate with an external OIDC provider. I think a lot of developers will equate the "Client" with stuff that runs in the browser particularly if they have some exposure to Blazor's ".Client" projects.
> 
> I was thinking the sample could be more similar to our Blazor OIDC sample at https://learn.microsoft.com/aspnet/core/blazor/security/blazor-web-app-with-oidc, but without the complexity of dealing with Blazor and the Auto render mode. I also think we could leave out the bits about how to configure Entra (since we'd recommend using AddMicrosoftIdentityWebApp for that anyway which we already have docs for), and we could leave figuring out the correct authority, client ID, and client secret as an exercise for the reader.

Thanks for the feedback. I updated the sample and the text and described. 
- Removed the IdentityServer
- Renamed the client
- Add the helper text to the sample

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+        {
+            RedirectUri = "/SignedOut"
+        },
+        CookieAuthenticationDefaults.AuthenticationScheme,
+        OpenIdConnectDefaults.AuthenticationScheme);
```
```suggestion
        // Clear auth cookie
        CookieAuthenticationDefaults.AuthenticationScheme,
        // Redirect to OIDC provider signout endpoint
        OpenIdConnectDefaults.AuthenticationScheme);
```

damienbod replied:
fixed

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+else
+{
+	<li class="nav-item">
+		<a class="nav-link text-dark" asp-area="" asp-page="/Index">Login</a>
```
Do we think it's worth showing how to initiate a challenge explicitly using `return Challenge()` or `HttpContext.ChallengeAsync()` similar to what we have for signout?

damienbod replied:
fixed, added a Login page using the HttpContext.ChallengeAsync()

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+    // .NET 9 feature
+    options.PushedAuthorizationBehavior = PushedAuthorizationBehavior.Require;
+
+    options.TokenValidationParameters.NameClaimType = JwtRegisteredClaimNames.Name;
+    options.TokenValidationParameters.RoleClaimType = "roles";
```
Isn't "role" more common? I also don't think the constant adds much value.

```suggestion
    options.TokenValidationParameters.NameClaimType = "name";
    options.TokenValidationParameters.RoleClaimType = "role";
```

damienbod replied:
fixed

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
````diff
@@ -0,0 +1,317 @@
+    options.TokenValidationParameters.RoleClaimType = "roles";
+});
+```
+
+See the following code example: 
+
+https://github.com/damienbod/keycloak-backchannel/tree/main/RazorPagePar
````
```suggestion
```

Does that sample show anything relevant to this doc that's not already in the sample we're adding with this PR? If so, I think we should just add it to our sample. Then it's less likely that the link goes stale too.

damienbod replied:
removed

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+
+See the [draft OAuth 2.0 for Browser-Based Applications](https://datatracker.ietf.org/doc/draft-ietf-oauth-browser-based-apps/) for further details.
+
+If implementing **web** applications which have no backend, it is recommended to use the [Backend for Frontend (BFF) pattern](/azure/architecture/patterns/backends-for-frontends) security architecture.  This pattern can be implemented in different ways, but the authentication is always implemented in the backend and no sensitive data is sent to the web client for further authorization or authenti ...
```
> which have no backend ... , but the authentication is always implemented in the backend

I'm not sure I follow what this is trying to say. It appears it's claiming that there both is and isn't a backend. I think we could clarify this by saying "which have no independent backend" or something like that.

damienbod replied:
fixed

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+
+### Logging
+
+Debugging OpenID Connect clients can be hard. Personally identifiable information (PII) data is not logged by default. If debugging in development mode, the ** IdentityModelEventSource.ShowPII** can be used to log sensitive personal data. This should never by deployed to productive servers.
```
This is a good callout. Can we also link to https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-8.0#configure-logging and mention that you may want to lower the configured log level to see everything?

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+    options.Scope.Add("offline_access");
+
+    options.ClaimActions.Remove("amr");
+    options.ClaimActions.MapJsonKey("website", "website");
```
```suggestion
    options.ClaimActions.MapUniqueJsonKey("website", "website");
```

I wonder if it would be good to put the `ClaimActions` stuff below `GetClaimsFromUserInfoEndpoint = true` to indicate they are related. It might also be good to link to https://learn.microsoft.com/en-us/aspnet/core/security/authentication/claims?view=aspnetcore-8.0#mapping-claims-using-openid-connect-authentication

halter73 commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+@if (Context.User.Identity!.IsAuthenticated)
+{
+	<li class="nav-item">
+		<a class="nav-link text-dark" asp-area="" asp-page="/Logout">Logout</a>
```
@blowdart Do we care about protecting logout endpoints with anti-csrf tokens?

halter73 reviewed (commented):
Thanks! This will be really nice to have

damienbod commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+
+### Setup the configuration properties
+
+Add the OpenID Connect client settings to the application configuration properties. The settings must match the client configuration in the OpenID Connect server. No secrets are persisted in the application settings, the secrets are stored in a Key Vault in production environments or in user secrets in a development environment.
```
fixed

damienbod commented on aspnetcore/security/authentication/configure-oidc-web-authentication.md:
```diff
@@ -0,0 +1,317 @@
+    options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+    options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+})
+.AddCookie()
```
Updated added a text for this

halter73 reviewed (approved):
Thank you!
